src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsProjectJsonToPackageReferenceMigrator.cs

```diff
@@ -23,34 +22,29 @@ internal class VsProjectJsonToPackageReferenceMigrator : IVsProjectJsonToPackage
         public VsProjectJsonToPackageReferenceMigrator(Lazy<IVsSolutionManager> solutionManager)
         {
             _solutionManager = solutionManager;
-            if(solutionManager == null)
+            if (solutionManager == null)
             {
                 throw new ArgumentNullException(nameof(solutionManager));
             }
         }
-        public IVsProjectJsonToPackageReferenceMigrateResult MigrateProjectJsonToPackageReference(string projectUniqueName)
+        public Task<object> MigrateProjectJsonToPackageReferenceAsync(string projectUniqueName)
         {
-            ThreadHelper.ThrowIfNotOnUIThread();
-            if(string.IsNullOrEmpty(projectUniqueName))
+            if (string.IsNullOrEmpty(projectUniqueName))
             {
                 throw new ArgumentNullException(nameof(projectUniqueName));
             }
 
-            if(!File.Exists(projectUniqueName))
+            if (!File.Exists(projectUniqueName))
             {
-                throw new FileNotFoundException(nameof(projectUniqueName));
+                throw new FileNotFoundException(string.Format(VsResources.Error_FileNotExists, projectUniqueName));
             }
 
-            return  NuGetUIThreadHelper.JoinableTaskFactory.Run(async () => 
-            {
-                await TaskScheduler.Default;
-                return await MigrateProjectToPackageRefAsync(projectUniqueName);
-            });
-            
+            return MigrateProjectToPackageRefAsync(projectUniqueName);
         }
 
-        private async Task<IVsProjectJsonToPackageReferenceMigrateResult> MigrateProjectToPackageRefAsync(string projectUniqueName)
-        {            
+        private async Task<object> MigrateProjectToPackageRefAsync(string projectUniqueName)
+        {
+            await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
```

Comment:
I disagree. The very next line in this method is `GetDTEProject` which will need the UI thread. By switching now, from an async method, we can do so without blocking a threadpool thread. It's best to do it here.

If there is a significant section of a method that doesn't need to be on the UI thread and you want to switch off for it, you can always do so for that section.